dotCMS/src/main/java/com/dotmarketing/factories/PublishFactory.java

```diff
@@ -288,20 +288,15 @@ public static boolean publishAsset(Inode webAsset, User user, boolean respectFro
 		}
 
 		if (webAsset instanceof Link) {
-			List contentlets = InodeFactory.getParentsOfClass(webAsset, com.dotmarketing.portlets.contentlet.business.Contentlet.class);
-			Iterator it = contentlets.iterator();
-			while (it.hasNext()) {
-				com.dotmarketing.portlets.contentlet.business.Contentlet cont = (com.dotmarketing.portlets.contentlet.business.Contentlet) it.next();
-			    if (cont.isLive()) {
-			    	try {
-			    		com.dotmarketing.portlets.contentlet.model.Contentlet newFormatContentlet =
-							conAPI.convertFatContentletToContentlet(cont);
-			    		    new ContentletLoader().invalidate(newFormatContentlet);
-					} catch (DotDataException e) {
-						throw new WebAssetException(e.getMessage(), e);
-					}
-			    }
-			}
+            FactoryLocator.getMenuLinkFactory()
+                    .getParentContentlets(webAsset.getInode()).stream().filter(contentlet -> {
+                try {
+                    return contentlet.isLive();
+                } catch (DotDataException | DotSecurityException  e) {
+                    throw new DotRuntimeException(e);
+                }
+			}).forEach( contentlet -> new ContentletLoader().invalidate(contentlet));
```

Comment:
I would create only one contentletLoader and reuse instead of create multiple copies of the instance, would save some memory if there is many contentlets to process

Reply:
Done